## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

/tmp/ipykernel_976146/2050583822.py:14: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)
/tmp/ipykernel_976146/2050583822.py:14: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [4]:
# 1a YOUR CODE HERE
contrib.shape

(1125659, 18)

- **1a answer here: It appears that we have 18 columns and 1.1 Million rows of data, this seems to match the 18 fields listed in the data file** 

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [59]:
# 1b YOUR CODE HERE
display(contrib.columns)

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')

- **1b answer here: yes, the 18 column names match what is in the documentation, with the exception of them being all lower case in this dataframe and CAPS in the document** 

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [6]:
# 1c YOUR CODE HERE
contrib.head(5)

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here: These 3 columns appear to represent Candidate ID, Candidate Name, and Contributor State.  They all appear to be in a string format** 

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [7]:
# 1d YOUR CODE HERE
contrib['election_tp']

0          P2016
1          P2016
2          P2016
3          P2016
4          P2016
           ...  
1125654    P2016
1125655    P2016
1125656    P2016
1125657    P2016
1125658    P2016
Name: election_tp, Length: 1125659, dtype: object

- **1d answer here: election_tp in the documentation is supposed to be the "Election Type/ Primary General Indicator".  This appears to be a code in the format EYYYY, where 'E' is the election type (Primary, General, Other, Convention, Runoff, Special, or Recount), followed by a 4 digit year as in '2016'.  'P2016' therefore means a primary election from the year 2016.** 

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [11]:
# 1e YOUR CODE HERE
print(contrib.dtypes)

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object


- **1e answer here: contrib_zip is of the type 'object' which is a generic type meaning there is likely mixed values in this column.  contb_receipt_amt is of the type float64 which means all objects are of a consistent type float.  contb_receipt_dt is of the type datetime64, meaning these are all of a consistent date - time format** 

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [13]:
# 1f YOUR CODE HERE
contrib.isnull().sum()

cmte_id                    0
cand_id                    0
cand_nm                    0
contbr_nm                  0
contbr_city               26
contbr_st                  0
contbr_zip                95
contbr_employer       157902
contbr_occupation      10399
contb_receipt_amt          0
contb_receipt_dt           0
receipt_desc         1110614
memo_cd               981391
memo_text             624511
form_tp                    0
file_num                   0
tran_id                    0
election_tp             1425
dtype: int64

- **1f answer here: The most nulls appear in the receipt_desc or 'Receipt Description' column.  There are 1.1106M nulls out of 1.1256M lines, indicating only 12% of the fields are populated.  If this field is not needed, it could be dropped. The next highest null count is memo_cd or 'Memo Code', with 981,391 nulls or  13% populated - this could also be dropped if it is not needed.** 

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [19]:
# 1g YOUR CODE HERE
contrib.groupby('cand_id')['cand_nm'].nunique()

cand_id
P00003392    1
P20002671    1
P20002721    1
P20003281    1
P20003984    1
P40003576    1
P60003670    1
P60005915    1
P60006046    1
P60006111    1
P60006723    1
P60007168    1
P60007242    1
P60007572    1
P60007671    1
P60007697    1
P60008059    1
P60008398    1
P60008521    1
P60008885    1
P60009685    1
P60022654    1
P80001571    1
P80003379    1
P80003478    1
Name: cand_nm, dtype: int64

- **1g answer here: The above code uses the groupby() and nunique() methods which tells the number of unique candidate names per candidate ID.  Because each item maps to exactly 1 unqiue name, we can conclude that every candidate ID is perfectly mapped to exactly 1 candidate name, including Sanders. ** 

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [27]:
# 1h YOUR CODE HERE
print(contrib['contbr_st'].nunique())
contrib['contbr_st']

1


0          CA
1          CA
2          CA
3          CA
4          CA
           ..
1125654    CA
1125655    CA
1125656    CA
1125657    CA
1125658    CA
Name: contbr_st, Length: 1125659, dtype: object

- **1h answer here: The two lines above prove that CA is the only state that was listed in this data.  There is only one unique field entered in the data, and by viewing this column we can see that ever field has CA entered.** 

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [46]:
# 1i YOUR CODE HERE
print(contrib['tran_id'].duplicated().sum())
contrib[contrib['tran_id'].duplicated(keep = False)].sort_values(by= 'tran_id')  # Sorts by tran_ID.  "keep = False" makes sure i see every occurence

3454


cmte_id    cand_id                    cand_nm             contbr_nm  \
42386  C00575449  P40003576                 Paul, Rand  PARKS, PATRICIA MRS.   
42835  C00575449  P40003576                 Paul, Rand  PARKS, PATRICIA MRS.   
43573  C00575449  P40003576                 Paul, Rand  PARKS, PATRICIA MRS.   
42428  C00575449  P40003576                 Paul, Rand       BREED, PAUL MR.   
42820  C00575449  P40003576                 Paul, Rand       BREED, PAUL MR.   
...          ...        ...                        ...                   ...   
15532  C00574624  P60006111  Cruz, Rafael Edward 'Ted'         BARTELS, DAVE   
36137  C00458844  P60006723               Rubio, Marco          RUBY, HOWARD   
37477  C00573519  P60005915        Carson, Benjamin S.        SAMPSON, RALPH   
36139  C00458844  P60006723               Rubio, Marco           RUSS, SUNNY   
37470  C00573519  P60005915        Carson, Benjamin S.       LEACH, FLORENCE   

         contbr_city contbr_st contbr_zip contbr_employer contbr_occupation  \
42386        HUGHSON        CA  95,326.00            SELF         HOUSEWIFE   
42835        HUGHSON        CA  95,326.00            SELF         HOUSEWIFE   
43573        HUGHSON        CA  95,326.00            SELF         HOUSEWIFE   
42428   SOLANA BEACH        CA  92,075.00      NET BURNER          ENGINEER   
42820   SOLANA BEACH        CA  92,075.00      NET BURNER          ENGINEER   
...              ...       ...        ...             ...               ...   
15532   LAGUNA HILLS        CA  92,653.00             NaN               NaN   
36137    LOS ANGELES        CA  90,064.00             NaN               NaN   
37477       CARLSBAD        CA  92,008.00             NaN               NaN   
36139  BEVERLY HILLS        CA  90,210.00             NaN               NaN   
37470         AUBURN        CA  95,602.00             NaN               NaN   

       contb_receipt_amt contb_receipt_dt receipt_desc memo_cd memo_text  \
42386             200.00       2015-11-28          NaN       X       NaN   
42835             200.00       2015-11-28          NaN       X       NaN   
43573             200.00       2015-11-28          NaN       X       NaN   
42428            -500.00       2015-10-28          NaN       X       NaN   
42820            -500.00       2015-10-28          NaN       X       NaN   
...                  ...              ...          ...     ...       ...   
15532          -2,700.00       2016-06-30       Refund     NaN       NaN   
36137          -1,800.00       2016-05-13       Refund     NaN       NaN   
37477            -100.00       2016-05-19       Refund     NaN       NaN   
36139            -200.00       2016-05-13       Refund     NaN       NaN   
37470             -50.00       2016-05-19       Refund     NaN       NaN   

      form_tp  file_num               tran_id election_tp  
42386   SA17A   1057796  A26C35A6066754130B99       P2016  
42835   SA17A   1057799  A26C35A6066754130B99       P2016  
43573   SA17A   1057798  A26C35A6066754130B99       P2016  
42428   SA17A   1057796  A340DF85B7F884133A20       P2016  
42820   SA17A   1057799  A340DF85B7F884133A20       P2016  
...       ...       ...                   ...         ...  
15532   SB28A   1096256           SB28A.90929         NaN  
36137   SB28A   1079053            SB28A.9119       G2016  
37477   SB28A   1079198            SB28A.9119         NaN  
36139   SB28A   1079053            SB28A.9127       G2016  
37470   SB28A   1079198            SB28A.9127         NaN  

[6873 rows x 18 columns]

- **1i answer here: By summing the duplicated ID's we can see that there are a total of 3454 duplicate items (Not including the original).  By viewing these duplicates,  it seems that a lot of them were on the same day by the same contributor.  A possible explanation could be individuals who wanted to donate to more than one candidate, or decided to donate to the same candidate more than once.** 

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [47]:
# 1j YOUR CODE HERE
contrib[contrib['contb_receipt_amt'] < 0]

cmte_id    cand_id                    cand_nm  \
19       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
23       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
81       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
190      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
213      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
...            ...        ...                        ...   
1125008  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125317  C00580399  P60008521   Christie, Christopher J.   
1125427  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125446  C00573519  P60005915        Carson, Benjamin S.   
1125582  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   

                    contbr_nm             contbr_city contbr_st  \
19           JOLLIFF, RICHARD                   CHICO        CA   
23           JOLLIFF, RICHARD                   CHICO        CA   
81           JOLLIFF, RICHARD                   CHICO        CA   
190          NOWELL, DIANA L.  RANCHO SANTA MARGARITA        CA   
213        LICHTY, ANDREW MR.               SAN DIEGO        CA   
...                       ...                     ...       ...   
1125008  NELSON, PETER C. MR.         SAN LUIS OBISPO        CA   
1125317       ASCHER, STEPHEN                PASADENA        CA   
1125427       HANSEN, WILLIAM                GLENDALE        CA   
1125446        PECK, JOHN JR.         RANCHO SANTA FE        CA   
1125582        FARRELL, PETER                LA JOLLA        CA   

            contbr_zip                     contbr_employer  \
19      959,289,507.00                       SELF EMPLOYED   
23      959,289,507.00                       SELF EMPLOYED   
81      959,289,507.00                       SELF EMPLOYED   
190     926,884,928.00  CAPISTRAND UNIFIED SCHOOL DISTRICT   
213     921,096,720.00                       SELF EMPLOYED   
...                ...                                 ...   
1125008      934018000                       SELF EMPLOYED   
1125317      911013113                             MIRAMAR   
1125427      912081507                             RETIRED   
1125446      920670829                    PECK ENTERPRISES   
1125582      920375633                              RESMED   

            contbr_occupation  contb_receipt_amt contb_receipt_dt  \
19                    RANCHER             -25.00       2016-04-29   
23                    RANCHER            -150.00       2016-04-29   
81                    RANCHER             -60.00       2016-04-14   
190      LIBRARIAN TECHNICIAN            -100.00       2016-04-11   
213               REAL ESTATE             -25.00       2016-04-30   
...                       ...                ...              ...   
1125008               DENTIST          -2,700.00       2015-07-31   
1125317             EXECUTIVE          -2,700.00       2016-01-30   
1125427               RETIRED          -5,400.00       2015-07-01   
1125446             PRESIDENT          -2,700.00       2016-01-01   
1125582              CHAIRMAN          -2,700.00       2015-08-28   

                     receipt_desc memo_cd                 memo_text form_tp  \
19       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
23       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
81       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
190      REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
213      REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
...                           ...     ...                       ...     ...   
1125008  REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
1125317   REATTRIBUTION TO SPOUSE       X   REATTRIBUTION TO SPOUSE   SA17A   
1125427                       NaN     NaN              CHARGED BACK   SA17A   
1125446  REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
1125582  REDESIGNATION TO GENERAL       X 

- **1j answer here: There are 11,896 lines that have a negative contribut9ion amount.  When look at the Receipt Description Column, it appears that most of these are labeled "REDESIGNATION TO GENERAL" or "REATTRIBUTION TO SPOUSE".  A likely explanation for this is that the funds that were initially deposited were reattributed to either a different candidate, election type, or donor.**

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [58]:
# 1k YOUR CODE HERE
print("Entries that were after the end date (June 7, 2016):")

display(contrib[contrib['contb_receipt_dt'] > '2017-06-07'])  # Greater than end date - No Results

print("Entries that were before the start date (Jan 1, 2014):")
display(contrib[contrib['contb_receipt_dt'] < '2014-01-01'])  # Less than start date

print("Entries that were before the start date (Jan 1, 2014):")
display(contrib[contrib['contb_receipt_dt'] < '2014-01-01'])  # Less than start date

print("Checking to see the number of unique entry lengths for date:")
display(contrib['contb_receipt_dt'].apply(lambda x: len(str(x))).nunique())

Entries that were after the end date (June 7, 2016):


Empty DataFrame
Columns: [cmte_id, cand_id, cand_nm, contbr_nm, contbr_city, contbr_st, contbr_zip, contbr_employer, contbr_occupation, contb_receipt_amt, contb_receipt_dt, receipt_desc, memo_cd, memo_text, form_tp, file_num, tran_id, election_tp]
Index: []

Entries that were before the start date (Jan 1, 2014):


cmte_id    cand_id       cand_nm          contbr_nm contbr_city  \
9932  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   
9994  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   

     contbr_st     contbr_zip contbr_employer  contbr_occupation  \
9932        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   
9994        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   

      contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
9932             -20.00       2013-11-05          NaN       X   
9994              20.00       2013-11-05          NaN       X   

                        memo_text form_tp  file_num             tran_id  \
9932  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.2.0615   
9994  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.3.0615   

     election_tp  
9932       P2016  
9994       G2016

Entries that were before the start date (Jan 1, 2014):


cmte_id    cand_id       cand_nm          contbr_nm contbr_city  \
9932  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   
9994  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   

     contbr_st     contbr_zip contbr_employer  contbr_occupation  \
9932        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   
9994        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   

      contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
9932             -20.00       2013-11-05          NaN       X   
9994              20.00       2013-11-05          NaN       X   

                        memo_text form_tp  file_num             tran_id  \
9932  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.2.0615   
9994  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.3.0615   

     election_tp  
9932       P2016  
9994       G2016

Checking to see the number of unique entry lengths for date:


1

- **1k answer here:  It appears that there are 2 results that came in 2013 before the start date, and no results after the end date.  It also appears that there the length of the string for receipt date is the same for all entries.  After looking through the data, it appears that there is a consistent and uniform format of yyy-mm-dd for every entry**

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here: yes, 18 columns and a little over 1.2M rows*

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:  Yes, we are interested in data from the 2016 Primary per zip code and funding per day of the month - all of these are present in this data set.**

**1l.3** What columns are important?

- **1l.3 answer here: The most important columsn to answer Bossman's questions are: contrb_zip, cont_receipt_amt, cand_id or cand_nm, contb_receipt_dt, and election_tp.  These 5-6 columns should be sufficient to answer the questions.** 

**1l.4** What columns can be dropped?

- **1l.4 answer here: "Receipt Description" and "Memo Code" can be dropped because of over 85% of the fields are NaN and it is not relevant to this analysis.  Additional fields that could be dropped that are not relevant are "form_tp", "file_num", "contbr_occupation", "contbr_employer", "contbr_nm", and "cmte_id".  Because the file size is not excesssively large, however, I would not drop these fields until I am CERTAIN that I do not need them.  There are usually follow up questions with this type of analysis, so it is good to have more information in your back pocket for when the eventual ask comes in - especially if memory and compute time are not overly compromised.** 

**1l.5** What are the data problems?

- **1l.5 answer here:  For this problem, we need to first filter by election type and candidate and then sum contributions based on Zip code and day of the month.  We then likely need to plot this information in a clean way to communicate the findings.**

**1l.6** List any assumptions so far:

- **1l.6 answer here:  For this straightforward analysis, we are assuming that the data in this file is accurate, that any duplicate fields are negligible, and that NaN fields can be ignored.**

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [84]:
# 2a YOUR CODE HERE
print("After filtering out anything not from the 2016 Primary we are left with the following dataframe shape: ")
contrib_P2016 = contrib[contrib['election_tp'] == 'P2016']
contrib_P2016.shape

After filtering out anything not from the 2016 Primary we are left with the following dataframe shape: 


(810481, 18)

**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [122]:
# 2b YOUR CODE HERE
bernie_contrib = contrib_P2016[contrib_P2016['cand_id'].str.contains('P60007168', na = False)]
bernie_contrib.shape

(407171, 18)

**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [127]:
# 2c YOUR CODE HERE

CA_zips = list(range(90001, 96163))  # https://www.maptive.com/interactive-maps/california-zip-codes/#:~:text=California%20has%20over%201%2C700%20ZIP,income%20details%20for%20every%20city.

#  https://stackoverflow.com/questions/21415661/logical-operators-for-boolean-indexing-in-pandas

str_zips_bool = "|".join( [str(x) for x in CA_zips] )

# zip_strs = [str(x) for x in CA_zips]

bernie_CA = bernie_contrib[bernie_contrib['contbr_zip'].str.contains(str_zips_bool, na = False)]  # Filters based on a Boolean String match in the ZIPS above
bernie_CA['clean_zips'] = bernie_CA['contbr_zip'].astype(str).str[:5]  #  Creates new column from first 5 items in original zip column.
bernie_CA


/tmp/ipykernel_976146/3726066128.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bernie_CA['clean_zips'] = bernie_CA['contbr_zip'].astype(str).str[:5]  #  Creates new column from first 5 items in original zip column.


cmte_id    cand_id           cand_nm          contbr_nm  \
240572  C00577130  P60007168  Sanders, Bernard       YOUNG, JASON   
240575  C00577130  P60007168  Sanders, Bernard        ROMO, SUSAN   
240578  C00577130  P60007168  Sanders, Bernard     ROSENBERG, LOU   
240579  C00577130  P60007168  Sanders, Bernard  CHITTENDEN, DAVID   
240580  C00577130  P60007168  Sanders, Bernard      TEWS, KURTISS   
...           ...        ...               ...                ...   
819180  C00577130  P60007168  Sanders, Bernard    PHILLIPS, VICKI   
819181  C00577130  P60007168  Sanders, Bernard  SCHAFFER, MICHAEL   
819182  C00577130  P60007168  Sanders, Bernard   SCIARONI, DANIEL   
819187  C00577130  P60007168  Sanders, Bernard  MATTHEWS, ADRIANA   
819195  C00577130  P60007168  Sanders, Bernard     DUFFY, CRELLAN   

          contbr_city contbr_st contbr_zip          contbr_employer  \
240572        ANAHEIM        CA  928064346       CRESCENT BAY FILMS   
240575    LOS ANGELES        CA  900271333    INFORMATION REQUESTED   
240578    LOS ANGELES        CA  900654038                     NONE   
240579    MILL VALLEY        CA  949411424            SELF EMPLOYED   
240580      SAN DIEGO        CA  921388620             NOT EMPLOYED   
...               ...       ...        ...                      ...   
819180  CARMEL VALLEY        CA  939249231             NOT EMPLOYED   
819181      SAN DIEGO        CA  921311618                     SELF   
819182   SHERMAN OAKS        CA  914113730  CULVER CITY URGENT CARE   
819187       EL CAJON        CA  920207369                  BRIXTON   
819195         NOVATO        CA  949496635            SELF EMPLOYED   

                contbr_occupation  contb_receipt_amt contb_receipt_dt  \
240572             VIDEO PRODUCER             100.00       2015-09-19   
240575      INFORMATION REQUESTED             232.62       2015-09-21   
240578               NOT EMPLOYED              10.00       2015-09-21   
240579                       SELF             250.00       2015-09-22   
240580               NOT EMPLOYED              10.00       2015-09-12   
...                           ...                ...              ...   
819180               NOT EMPLOYED              15.00       2016-03-20   
819181  CONSULTANT, FUSION ENERGY              50.00       2016-03-19   
819182                  PHYSICIAN              15.00       2016-03-18   
819187                         PR              14.00       2016-04-30   
819195                  ARCHITECT              27.00       2016-03-01   

       receipt_desc memo_cd                            memo_text form_tp  \
240572          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
240575          NaN     NaN                                  NaN   SA17A   
240578          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
240579          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
240580          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
...             ...     ...                                  ...     ...   
819180          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819181          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819182          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819187          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819195          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   

        file_num      tran_id election_tp clean_zips  
240572   1029414  VPF7BFJ41T1       P2016      92806  
240575   1029414  VPF7BFPMK32       P2016      90027  
240578   1029414  VPF7BFJGAW3       P2016      90065  
240579   1029414  VPF7BFK1452       P2016      94941  
240580   1029414  VPF7BFDCQV9       P2016      92138  
...          ...          ...         ...        ...  
819180   1077404  VPF7BMNPJD4       P2016      93924  
819181   1077404  VPF7BMN21E8       P2016      92131  
819182   1077404  VPF7BMMV4Q1       P2016  

**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [133]:
# 2d YOUR CODE HERE
bernie_CA_pos = bernie_CA[bernie_CA['contb_receipt_amt']>0]
bernie_CA_pos

cmte_id    cand_id           cand_nm          contbr_nm  \
240572  C00577130  P60007168  Sanders, Bernard       YOUNG, JASON   
240575  C00577130  P60007168  Sanders, Bernard        ROMO, SUSAN   
240578  C00577130  P60007168  Sanders, Bernard     ROSENBERG, LOU   
240579  C00577130  P60007168  Sanders, Bernard  CHITTENDEN, DAVID   
240580  C00577130  P60007168  Sanders, Bernard      TEWS, KURTISS   
...           ...        ...               ...                ...   
819180  C00577130  P60007168  Sanders, Bernard    PHILLIPS, VICKI   
819181  C00577130  P60007168  Sanders, Bernard  SCHAFFER, MICHAEL   
819182  C00577130  P60007168  Sanders, Bernard   SCIARONI, DANIEL   
819187  C00577130  P60007168  Sanders, Bernard  MATTHEWS, ADRIANA   
819195  C00577130  P60007168  Sanders, Bernard     DUFFY, CRELLAN   

          contbr_city contbr_st contbr_zip          contbr_employer  \
240572        ANAHEIM        CA  928064346       CRESCENT BAY FILMS   
240575    LOS ANGELES        CA  900271333    INFORMATION REQUESTED   
240578    LOS ANGELES        CA  900654038                     NONE   
240579    MILL VALLEY        CA  949411424            SELF EMPLOYED   
240580      SAN DIEGO        CA  921388620             NOT EMPLOYED   
...               ...       ...        ...                      ...   
819180  CARMEL VALLEY        CA  939249231             NOT EMPLOYED   
819181      SAN DIEGO        CA  921311618                     SELF   
819182   SHERMAN OAKS        CA  914113730  CULVER CITY URGENT CARE   
819187       EL CAJON        CA  920207369                  BRIXTON   
819195         NOVATO        CA  949496635            SELF EMPLOYED   

                contbr_occupation  contb_receipt_amt contb_receipt_dt  \
240572             VIDEO PRODUCER             100.00       2015-09-19   
240575      INFORMATION REQUESTED             232.62       2015-09-21   
240578               NOT EMPLOYED              10.00       2015-09-21   
240579                       SELF             250.00       2015-09-22   
240580               NOT EMPLOYED              10.00       2015-09-12   
...                           ...                ...              ...   
819180               NOT EMPLOYED              15.00       2016-03-20   
819181  CONSULTANT, FUSION ENERGY              50.00       2016-03-19   
819182                  PHYSICIAN              15.00       2016-03-18   
819187                         PR              14.00       2016-04-30   
819195                  ARCHITECT              27.00       2016-03-01   

       receipt_desc memo_cd                            memo_text form_tp  \
240572          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
240575          NaN     NaN                                  NaN   SA17A   
240578          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
240579          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
240580          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
...             ...     ...                                  ...     ...   
819180          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819181          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819182          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819187          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   
819195          NaN     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   

        file_num      tran_id election_tp clean_zips  
240572   1029414  VPF7BFJ41T1       P2016      92806  
240575   1029414  VPF7BFPMK32       P2016      90027  
240578   1029414  VPF7BFJGAW3       P2016      90065  
240579   1029414  VPF7BFK1452       P2016      94941  
240580   1029414  VPF7BFDCQV9       P2016      92138  
...          ...          ...         ...        ...  
819180   1077404  VPF7BMNPJD4       P2016      93924  
819181   1077404  VPF7BMN21E8       P2016      92131  
819182   1077404  VPF7BMMV4Q1       P2016  

**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [134]:
# 2e YOUR CODE HERE
bernie_CA_pos = bernie_CA_pos.drop(['receipt_desc', 'memo_cd', 'form_tp', 'file_num', 'contbr_employer', 'contbr_nm', 'cmte_id', 'memo_text', 'contbr_occupation'], axis=1)
bernie_CA_pos

cand_id           cand_nm    contbr_city contbr_st contbr_zip  \
240572  P60007168  Sanders, Bernard        ANAHEIM        CA  928064346   
240575  P60007168  Sanders, Bernard    LOS ANGELES        CA  900271333   
240578  P60007168  Sanders, Bernard    LOS ANGELES        CA  900654038   
240579  P60007168  Sanders, Bernard    MILL VALLEY        CA  949411424   
240580  P60007168  Sanders, Bernard      SAN DIEGO        CA  921388620   
...           ...               ...            ...       ...        ...   
819180  P60007168  Sanders, Bernard  CARMEL VALLEY        CA  939249231   
819181  P60007168  Sanders, Bernard      SAN DIEGO        CA  921311618   
819182  P60007168  Sanders, Bernard   SHERMAN OAKS        CA  914113730   
819187  P60007168  Sanders, Bernard       EL CAJON        CA  920207369   
819195  P60007168  Sanders, Bernard         NOVATO        CA  949496635   

        contb_receipt_amt contb_receipt_dt      tran_id election_tp clean_zips  
240572             100.00       2015-09-19  VPF7BFJ41T1       P2016      92806  
240575             232.62       2015-09-21  VPF7BFPMK32       P2016      90027  
240578              10.00       2015-09-21  VPF7BFJGAW3       P2016      90065  
240579             250.00       2015-09-22  VPF7BFK1452       P2016      94941  
240580              10.00       2015-09-12  VPF7BFDCQV9       P2016      92138  
...                   ...              ...          ...         ...        ...  
819180              15.00       2016-03-20  VPF7BMNPJD4       P2016      93924  
819181              50.00       2016-03-19  VPF7BMN21E8       P2016      92131  
819182              15.00       2016-03-18  VPF7BMMV4Q1       P2016      91411  
819187              14.00       2016-04-30  VPF7BPAHV45       P2016      92020  
819195              27.00       2016-03-01  VPF7BKS68R6       P2016      94949  

[21971 rows x 10 columns]

- **2e answer here:The following columns were dropped from the dataframe: ['receipt_desc', 'memo_cd', 'form_tp', 'file_num', 'contbr_employer', 'contbr_nm', 'cmte_id', 'memo_text', 'contbr_occupation']
These columns are not relevant to the analysis at hand.  Any column that pertained to election type, region, contribution, or candidate were kept.**

**2f.** List any assumptions that you made up to this point: 

- **2f answer here:  The assumptions made up until this point include all of the previous assumptions but now also include the following:

- the first 5 characters in the original zip code column, represent the base zip code in CA
- None of the columns that were dropped will be needed in this analysis
- All negative contribution amounts are negligble/ not needed for this analysis**

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [144]:
# 3a YOUR CODE HERE
print("the highest contribution amount by zip code: ")
display(bernie_CA_pos.groupby('clean_zips').sum('contb_receipt_amt').sort_values(by = 'contb_receipt_amt', ascending = False).head(5))

print("the highest contribution count by zip code: ")
display(bernie_CA_pos.groupby('clean_zips')['contb_receipt_amt'].count().sort_values(ascending = False).head(5))

the highest contribution amount by zip code: 


contb_receipt_amt
clean_zips                   
94110               21,953.46
94117               16,597.44
94114               15,879.72
94115               13,180.67
94122               11,306.04

the highest contribution count by zip code: 


clean_zips
94110    258
94114    199
94117    173
95060    145
95062    135
Name: contb_receipt_amt, dtype: int64

- **3a answer here:  Per the table above, the zip codes with the highest dollar amount of contributions are in order: 94110, 94117, 94114, 94115, and 94112

Per the table above, the zip codes with the highest count of contributions are in order: 94110, 94114, 94117, 95060, 95062

see above for dollar amounts and count.** 

**3b.** What day(s) of the month do most people donate? (10 points)

In [ ]:
# 3b YOUR CODE HERE
bernie_CA_pos['day'] = pd.to_datetime(bernie_CA_pos['contb_receipt_dt']).dt.day
bernie_CA_pos['day_name'] = pd.to_datetime(bernie_CA_pos['contb_receipt_dt']).dt.day_name()

day_summary = bernie_CA_pos.groupby(['day'])['contb_receipt_amt'].count().sort_values(ascending = False)

day_name_summary = bernie_CA_pos.groupby(['day_name'])['contb_receipt_amt'].count().sort_values(ascending = False)

print("The days of the month with the highest number of contributors are shown in the table below: ")  # Day of the Month
display(day_summary.head(10))

print("The days of the week with the highest number of contributors are shown in table below: ")  # Day of the Week
display(day_name_summary.head(10))

The days of the month with the highest number of contributors are shown in the table below: 


day
9     2887
31    2827
14    1872
27    1337
29    1060
8     1018
26     971
25     906
28     884
7      816
Name: contb_receipt_amt, dtype: int64

The days of the week with the highest number of contributors are shown in table below: 


day_name
Wednesday    4613
Monday       3766
Tuesday      3624
Thursday     3414
Sunday       2629
Saturday     2142
Friday       1783
Name: contb_receipt_amt, dtype: int64

- **3b answer here:  The days of the month (top 5) with the highest number of contributors in order are the following: 9th, 31st, 14th, 27th, and 29th.  See outputs above for number of contributors on these days.

The days of the week (top 5) with the highest number of contributors in order are the following: Wednesday, Monday, Tuesday, Thursday, and Sunday.  See outputs above for number of contributors on these days. ** 

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k